File: backend/src/services/kpi.service.js
import supabase from '../config/supabase.js';

export async function kpiGlobal() {
  // COUNT exacto
  const { count, error: countErr } = await supabase
    .from('respuestas_formulario')
    .select('*', { count: 'exact', head: true });
  if (countErr) throw countErr;

  // AVG de satisfaccion (simple para claridad)
  const { data, error } = await supabase
    .from('respuestas_formulario')
    .select('satisfaccion');
  if (error) throw error;

  const valores = (data || []).map(r => r.satisfaccion || 0);
  const promedio = valores.length ? (valores.reduce((a,b) => a+b, 0) / valores.length) : 0;

  return { total: count || 0, promedio: Number(promedio.toFixed(2)) };
}

export async function kpiPorCategoria() {
  const { data, error } = await supabase
    .from('respuestas_formulario')
    .select('categoria, satisfaccion');
  if (error) throw error;

  const agg = {};
  for (const r of (data || [])) {
    const cat = r.categoria || 'Sin categoría';
    if (!agg[cat]) agg[cat] = { total: 0, sum: 0 };
    agg[cat].total += 1;
    agg[cat].sum += r.satisfaccion || 0;
  }

  return Object.entries(agg).map(([categoria, v]) => ({
    categoria,
    total_respuestas: v.total,
    promedio_satisfaccion: Number((v.sum / v.total).toFixed(2))
  }));
}

File: backend_fastapi/app/services/kpi_service.py
from ..config import supabase

async def obtener_kpis():
    # COUNT exacto
    total_resp = supabase.table("respuestas_formulario").select("*", count="exact", head=True).execute()
    total = getattr(total_resp, "count", None) or 0

    # Traer satisfacciones para promedio
    rows = supabase.table("respuestas_formulario").select("satisfaccion, categoria").execute().data or []

    # Promedio global
    vals = [r.get("satisfaccion") or 0 for r in rows if r.get("satisfaccion") is not None]
    promedio = round(sum(vals) / len(vals), 2) if vals else 0.0

    # Agregación por categoría
    agg = {}
    for r in rows:
        cat = r.get("categoria") or "Sin categoría"
        sat = r.get("satisfaccion") or 0
        if cat not in agg:
            agg[cat] = {"total": 0, "sum": 0}
        agg[cat]["total"] += 1
        agg[cat]["sum"] += sat

    categorias = [
        {
            "categoria": cat,
            "total_respuestas": v["total"],
            "promedio_satisfaccion": round(v["sum"]/v["total"], 2) if v["total"] else 0.0
        }
        for cat, v in agg.items()
    ]

    return {
        "global": {"total": total, "promedio": promedio},
        "categorias": categorias
    }
